File: ClassDemo/w12-1_TreeCodeNotebooks/TreeCodeRelease/nbody.py
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d
import numpy as np
from treeclasses import *
from timerclass import Timer

class Hamiltonian:
    def __init__(self, theta: float, epsSmooth:float, maxLeafSize:int, maxSrcLen:int):
        self.theta = theta
        self.epsSmooth = epsSmooth
        self.maxLeafSize = maxLeafSize
        self.maxSrcLen = maxSrcLen

        self.verbose = False
        self.check = False
        self.getStats = False

    def getAcceleration(self, ps: ParticleSet):

        ps.computeBoundingBox()
        center, halfWidth = ps.getBoundingBox()

        BH = BHTree(ps, self.maxLeafSize, self.epsSmooth, self.maxSrcLen)

        BH.makeTree(self.verbose, self.check, self.getStats)

        BH.BHsubsets(theta, ps.N)
        #BH.accAll(theta)

        BH.putParticlesInOriginalOrder()  # so we know which particles are which galaxy

        BH.free()                         # important to free memory on the C side...

    def positionEquation(self, ps: ParticleSet, newvel: bool) -> np.ndarray :
        return ps.v

    def momentumEquation(self, ps: ParticleSet, newacc: bool) -> np.ndarray :
        if newacc: self.getAcceleration(ps)
        return ps.a


class State:
    def __init__(self, time: float, ps: ParticleSet, hamilton: Hamiltonian ):
        self.time = time
        self.step = 0
        self.ps = ps

    def kick(self, h, hamilton, recalculate):
        # Update velocity
        self.ps.v += h * hamilton.momentumEquation(self.ps, recalculate)
        return self

    def drift(self, h, hamilton, recalculate):
        # Update positions
        self.ps.r += h * hamilton.positionEquation(self.ps, recalculate)
        return self

def KDK(dt: float, hamilton: Hamiltonian, s: State):
    s = s.kick(dt/2, hamilton, False).drift(dt, hamilton, False).kick(dt/2, hamilton, True)
    s.time += dt
    s.step += 1
    return s

def getEnergy(ps: ParticleSet):
    K = 0.5*np.sum(ps.mass * np.sum(ps.v*ps.v, axis=1))
    P = 0.5*np.sum(ps.mass * ps.pot)
    mom = np.sum(ps.mass[:,np.newaxis] * ps.v)

    return K, P, mom

with open("dubinski.tab") as f:
    data = np.loadtxt(f)

# create a particle set
PS = ParticleSet()

# allocate storage
N = data.shape[0]
PS.reserve(N)

# copy in the data
PS.mass[:] = data[:,0]
PS.r[:,:]  = data[:,1:4]
PS.v[:,:]  = data[:,4:]

# the data consists of sets of particles for the disks, bulges, and halos in each galaxy
Ndisk = 16384
Nbulge = 8192
Nhalo = 16384
Ngal = Ndisk+Nbulge+Nhalo
assert( Ngal*2 == N )

disk1 = np.s_[:Ndisk+Nbulge]
disk2 = np.s_[Ngal:Ngal+Ndisk+Nbulge]

PS.computeBoundingBox()
centre, halfWidth = PS.getBoundingBox()

fig = plt.figure()
ax = plt.axes(projection='3d')
c, hw = PS.getBoundingBox()
ax.set_xlim(c[0]-hw,c[0]+hw)
ax.set_ylim(c[0]-hw,c[0]+hw)
ax.set_zlim(c[0]-hw,c[0]+hw)
ax.set_xlabel('x')
ax.set_ylabel('y')
ax.set_zlabel('z')

plot1, = ax.plot3D(PS.r[disk1,0], PS.r[disk1,1], PS.r[disk1,2], 'b.', alpha=0.05, markersize=0.5)
plot2, = ax.plot3D(PS.r[disk2,0], PS.r[disk2,1], PS.r[disk2,2], 'r.', alpha=0.05, markersize=0.5)
steptxt = ax.text2D(0.2,1.0, f"step:    0   time: {0.0:8.2e}\nerror: E: {0:.2e}  P: {0:.2e}", transform=ax.transAxes)

plt.show(block=False)

def updatePlot(S, timer):
    ps = S.ps
    plot1.set_data(ps.r[disk1,0], ps.r[disk1,1]);  plot1.set_3d_properties(ps.r[disk1,2])
    plot2.set_data(ps.r[disk2,0], ps.r[disk2,1]);  plot2.set_3d_properties(ps.r[disk2,2])

    K, P, mom = getEnergy(PS)
    DErel = (K+P-E0)/E0
    DP = np.max(np.abs(mom-mom0))
    line = f"step: {S.step:4d}   time: {S.time:8.2e}\nerror: E: {DErel:.2e}  P: {DP:.2e}\nrate: {ps.N/timer.elapsed():.2e}"
    steptxt.set_text(line)

    # this updates the animation w/o stealing the window focus
    plt.gcf().canvas.draw_idle(); plt.gcf().canvas.start_event_loop(0.001)

maxLeafSize = 16
maxSrcLen = PS.N
theta = 0.75
epsSmooth = 0.025
dt = 0.1
skip = 2

H = Hamiltonian(theta, epsSmooth, maxLeafSize, maxSrcLen)
S = State(0.0, PS, H)

S = KDK(0.0, H, S) # take a step w/ dt=0 just to get initial potential
K0, P0, mom0 = getEnergy(PS)
E0 = K0+P0


def runit():
    global S
    timer = Timer()

    while(S.time<1.0):

        timer.cstart()
        S = KDK(dt, H, S) # take a KDK step
        timer.stop()

        if S.step%skip==0:
            updatePlot(S, timer)

    updatePlot(S, timer)


# profile code
# run snakeviz on nbody.prof output
if __name__ == "__main__":
    import cProfile
    import pstats

    with cProfile.Profile() as pr:
        runit()

    stats = pstats.Stats(pr)
    stats.sort_stats('tottime')
    stats.dump_stats("nbody.pprof")

File: ClassDemo/w12-1_TreeCodeNotebooks/TreeCodeRelease/nbody.cc
/*
  Simple implementation of an N-body code entirely in C++.
 */



#include "STimer.cc"
#include "pprint.cc"
#include "util.cc"
#include "particleset.cc"
#include "bhtree.cc"

/*
  Hamiltonian class provides the update for the timestepper
 */
template <typename T>
class Hamiltonian {
public:

    using vec = SmallVec<T, 3>;

    Hamiltonian(double theta, int maxLeafSize, int maxSrcLen, int maxInteractionListLength, double epsSmooth)
        : theta(theta), maxLeafSize(maxLeafSize), maxSrcLen(maxSrcLen),
          maxInteractionListLength(maxInteractionListLength),
          epsSmooth(epsSmooth) {  }

    void getAcc(ParticleSet<T> &ps) {

        ps.computeBoundingBox();

         BHtree<T> BH = BHtree<T>(ps, maxLeafSize, epsSmooth, maxSrcLen);
        //BH->getStats = true;
        //BH->check = true;
        //BH->verbose = true;
        BH.makeTree();

        BH.BHsubsets(theta,  maxInteractionListLength, vec(0.0));
        BH.reportStats();
    }

    void update_v(ParticleSet<T> &ps) {
        // nothing to do
    }

    void update_a(ParticleSet<T> &ps) {
        // update acceleration from positions
        getAcc(ps);
    }

    //BHtree<T> *BH;

    int maxLeafSize, maxSrcLen, maxInteractionListLength;
    double theta, epsSmooth;

};

/*
  State holds the state of the system and implements the kick and drift operators
  using an instance of Hamiltonian.
 */
template <typename T>
class State {
public:
    State(ParticleSet<T> &ps, Hamiltonian<T> &H) : ps(ps), H(H), time(0) {}

    State *kick(double dt, bool recalc) {
        if(recalc) H.update_a(ps);
        for(int i=0; i<ps.N; i++) ps[i].v += dt * ps[i].a;
        return this;
    }

    State *drift(double dt, bool recalc) {
        if(recalc) H.update_v(ps);
        for(int i=0; i<ps.N; i++) ps[i].r += dt * ps[i].v;
        return this;
    }

    double time;
    int step=0;
    ParticleSet<T> &ps;
    Hamiltonian<T> &H;
};

/*
  step implements KDK or DKD leapfrog integrators.
*/
enum Method { KDK, DKD };
template <Method M, typename T>
void step(double dt, State<T> *s) {
    if constexpr (M == Method::KDK) {
            s = s->kick(0.5*dt,false)->drift(dt,false)->kick(0.5*dt,true);
            s->time += dt;
            s->step++;
        }
    if constexpr (M == Method::DKD) {
            s = s->drift(0.5*dt,false)->kick(dt,true)->drift(0.5*dt,false);
            s->time += dt;
            s->step++;
   }
}

/*
  getE computes the system totak kinetic and potential energy.
 */
template <typename T>
class Conserve {
public:
    Conserve(ParticleSet<T> &ps) : ps(ps) {
        getE(K0, P0, E0);
    }

    void getE(double &K, double &P, double &E) {
        K = P = 0;
        for(int i=0; i<ps.N; i++) {
            K += 0.5*ps[i].v.norm2()*ps[i].mass;
            P += ps[i].pot*ps[i].mass;
        }
        P *= 0.5;
        E = K+P;
    }

    void printCons() {
        double K, P, E;
        getE(K, P, E);
        pprint("E error: %.2e\n", (K+P-E0)/E0);
    }

    ParticleSet<T> &ps;
    double K0, P0, E0;
};

template <typename T>
void readDubinski(ParticleSet<T> &ps) {

    using vec = SmallVec<T,3>;

    std::string fname("dubinski.tab");

    // count number of lines in file
    std::ifstream f(fname,std::ios::in);
    int N = 0;
    std::string line;
    while(!f.eof()) { std::getline(f,line); N++; }
    f.close();
    N--;
    pprint("read %d lines from %s\n", N, fname);

    // read in data
    ps.reserve(N);
    f.open(fname,std::ios::in);
    for(int i=0; i<N; i++) {
        double m, x, y, z, vx, vy, vz;
        f >> m >> x >> y >> z >> vx >> vy >> vz;
        ps[i].mass = m;
        ps[i].r = vec(x,y,z);
        ps[i].v = vec(vx,vy,vz);
    }
    f.close();
}

int main() {

    using vec = SmallVec<double, 3>;

    int N = pow(2,20);
    ParticleSet<double> ps(0);
    readDubinski(ps);
    ps.computeBoundingBox();
    pprint("center: %e  halfWidth: %e\n\n", ps.centre, ps.halfWidth);

    double theta = 0.6;
    int maxLeafSize = 32;
    int maxSrcLen = ps.N;
    int maxInteractionListLength =5000;
    double epsSmooth = 0.98*pow(ps.N, -0.26) * ps.halfWidth;
    Hamiltonian<double> H(theta, maxLeafSize, maxSrcLen, maxInteractionListLength, epsSmooth);

    State s(ps, H);

    // zero-dt step to get potential
    step<KDK>(0.0, &s);
    Conserve C(ps);

    double dt = 0.1;

    STimer steptime;
    STimer fulltime;
    fulltime.Clear();
    while(s.time<20.0) {
        steptime.CStart();
        fulltime.Start();
        step<KDK>(dt, &s);
        steptime.Stop();
        fulltime.Stop();
        double rate = ps.N/steptime.Elapsed();
        if(s.step%10==0) { pprint("%5d rate: %e  time: %e  ", s.step, rate, fulltime.Elapsed()); C.printCons(); }
    }

}
